// @ui5/webcomponents-react/dist/components/ObjectPageSection/index.js
'use client';

import { useStylesheet } from '@ui5/webcomponents-react-base';
import { clsx } from 'clsx';
import React, { forwardRef } from 'react';
import { classNames, styleData } from './ObjectPageSection.module.css.js';
/**
 * Top-level information container of an `ObjectPage`.
 */
const ObjectPageSection = /*#__PURE__*/forwardRef((props, ref) => {
  const {
    titleText = '',
    id,
    children,
    titleTextUppercase = true,
    className,
    style,
    hideTitleText,
    titleTextLevel = 'H3',
    wrapTitleText,
    header,
    ...rest
  } = props;
  useStylesheet(styleData, ObjectPageSection.displayName);
  const htmlId = `ObjectPageSection-${id}`;
  const titleClasses = clsx(classNames.title, titleTextUppercase && classNames.uppercase);
  return /*#__PURE__*/React.createElement("section", {
    ref: ref,
    role: "region",
    className: clsx(classNames.section, wrapTitleText && classNames.wrap, className),
    style: style,
    ...rest,
    id: htmlId,
    "data-component-name": "ObjectPageSection"
  }, !!header && /*#__PURE__*/React.createElement("div", {
    className: classNames.headerContainer
  }, header), !hideTitleText && /*#__PURE__*/React.createElement("div", {
    role: "heading",
    "aria-level": parseInt(titleTextLevel.slice(1)),
    className: classNames.titleContainer,
    "data-component-name": "ObjectPageSectionTitleText"
  }, /*#__PURE__*/React.createElement("div", {
    className: titleClasses
  }, titleText)), /*#__PURE__*/React.createElement("div", {
    className: classNames.sectionContent
  }, /*#__PURE__*/React.createElement("div", {
    className: classNames.sectionContentInner,
    "data-component-name": "ObjectPageSectionContent"
  }, children)));
});
ObjectPageSection.displayName = 'ObjectPageSection';
export { ObjectPageSection };
