// @ui5/webcomponents-react/wrappers/components/AnalyticalTable/defaults/LoadingComponent/TablePlaceholder.js
import { ThemingParameters, useStylesheet } from '@ui5/webcomponents-react-base';
import { clsx } from 'clsx';
import React from 'react';
import { resolveCellAlignment } from '../../util/index.js';
import { classNames, styleData } from './TablePlaceholder.module.css.js';
const getArrayOfLength = len => Array.from(Array(len).keys());
export const TablePlaceholder = props => {
  const {
    columns,
    rows,
    style
  } = props;
  useStylesheet(styleData, TablePlaceholder.displayName);
  return /*#__PURE__*/React.createElement("div", {
    style: {
      backgroundColor: ThemingParameters.sapList_Background,
      width: '100%',
      ...style
    },
    "data-component-name": "AnalyticalTableLoadingPlaceholder"
  }, getArrayOfLength(rows).map((_, index) => {
    return /*#__PURE__*/React.createElement("div", {
      className: classNames.row,
      key: `row-${index}`
    }, columns.map(col => {
      return /*#__PURE__*/React.createElement("div", {
        key: `row${index}-${col.id}`,
        className: classNames.cellContainer,
        style: {
          width: col.totalWidth,
          ...resolveCellAlignment(col)
        }
      }, /*#__PURE__*/React.createElement("div", {
        className: clsx(classNames.cell, classNames.animation)
      }));
    }));
  }));
};
TablePlaceholder.displayName = 'TablePlaceholder';
TablePlaceholder.defaultProps = {
  rows: 5
};
