// @ui5/webcomponents-react/wrappers/components/DynamicPageHeader/index.js
'use client';

import { useStylesheet } from '@ui5/webcomponents-react-base';
import { clsx } from 'clsx';
import React, { forwardRef, useMemo } from 'react';
import { classNames, styleData } from './DynamicPageHeader.module.css.js';
/**
 * The `DynamicPageHeader` component is part of the `DynamicPage` family and is used to serve as header section of the `DynamicPage` and `ObjectPage`.
 * This component can be collapsed and pinned by the anchorbar.
 *
 * __Note:__ When used inside a custom component, it's essential to pass through all props, as otherwise the component won't function as intended!
 */
const DynamicPageHeader = /*#__PURE__*/forwardRef((props, ref) => {
  const {
    children,
    headerPinned,
    topHeaderHeight,
    className,
    style,
    ...rest
  } = props;
  const headerStyles = useMemo(() => {
    if (headerPinned) {
      return {
        ...style,
        top: `${topHeaderHeight}px`,
        zIndex: 1
      };
    }
    return style;
  }, [headerPinned, topHeaderHeight, style]);
  useStylesheet(styleData, DynamicPageHeader.displayName);
  return /*#__PURE__*/React.createElement("div", {
    ref: ref,
    ...rest,
    className: clsx(classNames.header, className),
    "data-component-name": "DynamicPageHeader",
    style: headerStyles
  }, children);
});
DynamicPageHeader.displayName = 'DynamicPageHeader';
export { DynamicPageHeader };
