// @ui5/webcomponents-react/dist/components/ObjectPageSubSection/index.js
'use client';

import { useStylesheet } from '@ui5/webcomponents-react-base';
import { clsx } from 'clsx';
import React, { forwardRef } from 'react';
import { FlexBoxAlignItems, FlexBoxDirection, FlexBoxJustifyContent } from '../../enums/index.js';
import { FlexBox } from '../FlexBox/index.js';
import { classNames, styleData } from './ObjectPageSubSection.module.css.js';
/**
 * Second-level information container of an `ObjectPage`.<br />
 * __Note:__ This component should only be used inside an `ObjectPageSection` component.
 */
const ObjectPageSubSection = /*#__PURE__*/forwardRef((props, ref) => {
  const {
    children,
    id,
    titleText,
    className,
    style,
    actions,
    hideTitleText,
    titleTextLevel = 'H4',
    titleTextUppercase,
    ...rest
  } = props;
  const htmlId = `ObjectPageSubSection-${id}`;
  useStylesheet(styleData, ObjectPageSubSection.displayName);
  const subSectionClassName = clsx(classNames.objectPageSubSection, className);
  return /*#__PURE__*/React.createElement("div", {
    ref: ref,
    role: "region",
    style: style,
    tabIndex: -1,
    ...rest,
    className: subSectionClassName,
    id: htmlId,
    "data-component-name": "ObjectPageSubSection"
  }, /*#__PURE__*/React.createElement(FlexBox, {
    direction: FlexBoxDirection.Row,
    justifyContent: FlexBoxJustifyContent.SpaceBetween,
    alignItems: FlexBoxAlignItems.Center,
    className: classNames.headerContainer,
    "data-component-name": "ObjectPageSubSectionHeaderContainer"
  }, !hideTitleText ? /*#__PURE__*/React.createElement("div", {
    role: "heading",
    "aria-level": parseInt(titleTextLevel.slice(1)),
    className: clsx(classNames.subSectionTitle, titleTextUppercase && classNames.uppercase),
    "data-component-name": "ObjectPageSubSectionTitleText"
  }, titleText) : /*#__PURE__*/React.createElement("span", {
    "aria-hidden": "true",
    className: classNames.spacer
  }), actions && actions), /*#__PURE__*/React.createElement("div", {
    className: classNames.subSectionContent,
    "data-component-name": "ObjectPageSubSectionContent"
  }, children));
});
ObjectPageSubSection.displayName = 'ObjectPageSubSection';
export { ObjectPageSubSection };
